File: packages/relay-compiler/transforms/query-generators/QueryQueryGenerator.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

// flowlint ambiguous-object-type:error

'use strict';

const {
  buildFragmentSpread,
  buildOperationArgumentDefinitions,
} = require('./utils');

import type {Fragment} from '../../core/IR';
import type {Schema} from '../../core/Schema';
import type {QueryGenerator, RefetchRoot} from '.';

function buildRefetchOperation(
  schema: Schema,
  fragment: Fragment,
  queryName: string,
): ?RefetchRoot {
  const queryType = schema.expectQueryType();
  if (!schema.areEqualTypes(fragment.type, queryType)) {
    return null;
  }

  const {directives} = fragment.metadata || {}

  const a = ([]: Array<Directive>)
  const args = ([]: Array<LocalArgumentDefinition>)
  if (directives != null) {
    a.push(...((directives:any): Array<Directive>))
    a.forEach(d=>{
      d.args.forEach(aa=>{
        if (aa.value.kind == 'Variable' && null != aa.value.variableName && null != aa.value.type) {
          args.push({
            defaultValue:   null,
            kind: 'LocalArgumentDefinition',
            loc: aa.value.loc,
            name: aa.value.variableName,
            type: aa.value.type,
          })
        }
      })
    })
  }

  return {
    identifierField: null,
    path: [],
    node: {
      argumentDefinitions: [...buildOperationArgumentDefinitions(
        fragment.argumentDefinitions,
      ), ...args],
      directives: [...a],
      kind: 'Root',
      loc: {kind: 'Derived', source: fragment.loc},
      metadata: null,
      name: queryName,
      operation: 'query',
      selections: [buildFragmentSpread(fragment)],
      type: queryType,
    },
    transformedFragment: fragment,
  };
}

module.exports = ({
  description: 'the Query type',
  buildRefetchOperation,
}: QueryGenerator);

File: compiler/crates/graphql-transforms/src/refetchable_fragment/query_query_generator.rs
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use super::{
    build_fragment_metadata_as_directive, build_fragment_spread,
    build_operation_variable_definitions, build_used_global_variables, QueryGenerator, RefetchRoot,
    RefetchableDerivedFromMetadata, RefetchableMetadata,
};
use crate::root_variables::VariableMap;
use common::{DiagnosticsResult, WithLocation};
use graphql_ir::{FragmentDefinition, OperationDefinition};
use graphql_syntax::OperationKind;
use interner::StringKey;
use schema::Schema;
use std::sync::Arc;

fn build_refetch_operation(
    schema: &Schema,
    fragment: &Arc<FragmentDefinition>,
    query_name: StringKey,
    variables_map: &VariableMap,
) -> DiagnosticsResult<Option<RefetchRoot>> {
    let query_type = schema.query_type().unwrap();
    if fragment.type_condition != query_type {
        return Ok(None);
    }

    let fragment = Arc::new(FragmentDefinition {
        directives: build_fragment_metadata_as_directive(
            fragment,
            RefetchableMetadata {
                operation_name: query_name,
                path: vec![],
                identifier_field: None,
            },
        ),
        used_global_variables: build_used_global_variables(
            variables_map,
            &fragment.variable_definitions,
        )?,
        variable_definitions: fragment.variable_definitions.clone(),
        ..fragment.as_ref().clone()
    });
    Ok(Some(RefetchRoot {
        operation: Arc::new(OperationDefinition {
            kind: OperationKind::Query,
            name: WithLocation::new(fragment.name.location, query_name),
            type_: query_type,
            variable_definitions: build_operation_variable_definitions(&fragment),
            directives: vec![RefetchableDerivedFromMetadata::create_directive(
                fragment.name,
            )],
            selections: vec![build_fragment_spread(&fragment)],
        }),
        fragment,
    }))
}

pub const QUERY_QUERY_GENERATOR: QueryGenerator = QueryGenerator {
    description: "the Query type",
    build_refetch_operation,
};
